rewrite-core/src/main/java/org/openrewrite/scheduling/RecipeRunCycle.java

```diff
@@ -112,7 +112,11 @@ public LSS generateSources(LSS sourceSet) {
                 ScanningRecipe<Object> scanningRecipe = (ScanningRecipe<Object>) recipe;
                 List<SourceFile> generated = new ArrayList<>(scanningRecipe.generate(scanningRecipe.getAccumulator(rootCursor, ctx), unmodifiableList(acc), ctx));
                 generated.replaceAll(source -> addRecipesThatMadeChanges(recipeStack, source));
-                acc.addAll(generated);
+                for (SourceFile source : generated) {
+                    if (acc.stream().noneMatch(s -> s.getSourcePath().equals(source.getSourcePath()))) {
+                        acc.add(source);
+                    }
+                }
```

Comment:
Gave me a bit of pause to loop over all generated files in a list before adding a new entry, but I think it's tolerable given the expected low volume of generated files.

Reply:
Change the reduction operation to use a `Set` generated from a `TreeMap` whose comparator is the source path.

Reply:
I updated to using `TreeSet`. This does mean that the insertion order is no longer guaranteed to be the resulting order, although I can imagine that the LST order is also similarly ordered anyway?